sentry-ruby/lib/sentry/profiler.rb

```diff
@@ -0,0 +1,232 @@
+# frozen_string_literal: true
+
+return unless defined?(StackProf)
+
+require 'securerandom'
+
+module Sentry
+  class Profiler
+
+    VERSION = '1'
+    PLATFORM = 'ruby'
+    # 101 Hz in microseconds
+    DEFAULT_INTERVAL = 1e6 / 101
+
+    attr_reader :sampled, :started, :event_id
+
+    def initialize
+      @event_id = SecureRandom.uuid.delete('-')
+      @started = false
+      @sampled = nil
+    end
+
+    def start
+      return unless @sampled
+
+      @started = StackProf.start(interval: DEFAULT_INTERVAL,
+                                 mode: :wall,
+                                 raw: true,
+                                 aggregate: false)
+
+      @started ? log('Started') : log('Not started since running elsewhere')
+    end
+
+    def stop
+      return unless @sampled
+      return unless @started
+
+      StackProf.stop
+      log('Stopped')
+    end
+
+    def profile_context
+      { profile_id: @event_id }
+    end
+
+    # Sets initial sampling decision of the profile.
+    # @return [void]
+    def set_initial_sample_decision(transaction_sampled)
+      unless Sentry.initialized? && Sentry.configuration.profiling_enabled?
+        @sampled = false
+        return
+      end
+
+      unless transaction_sampled
+        @sampled = false
+        log('Discarding profile because transaction not sampled')
+        return
+      end
+
+      profiles_sample_rate = Sentry.configuration.profiles_sample_rate
+
+      case profiles_sample_rate
+      when 0.0
+        @sampled = false
+        log('Discarding profile because sample_rate is 0')
+        return
+      when 1.0
+        @sampled = true
+        return
+      else
+        @sampled = Random.rand < profiles_sample_rate
+      end
+
+      log('Discarding profile due to sampling decision') unless @sampled
+    end
+
+    def to_hash
+      return nil unless Sentry.initialized?
+      return nil unless @sampled
+      return nil unless @started
+
+      results = StackProf.results
+      return nil unless results
+      return nil if results.empty?
+      return nil if results[:samples] == 0
+      return nil unless results[:raw]
+
+      frame_map = {}
+
+      frames = results[:frames].to_enum.with_index.map do |frame, idx|
+        frame_id, frame_data = frame
+
+        # need to map over stackprof frame ids to ours
+        frame_map[frame_id] = idx
+
+        in_app = in_app?(frame_data[:file])
```

Comment:
Can we declare `frame_data[:file]` as a local variable? Likely `file_path`?

Reply:
not sure I follow, this is for each `frame`, what would the local variable be?

Reply:
I mean there are 3 occurrences of `frame_data[:file]`, maybe we can extract it to a local variable?

https://github.com/getsentry/sentry-ruby/blob/b0e92692c9c451da6fc7da6810f5fd2a0f2a74b5/sentry-ruby/lib/sentry/profiler.rb#L100-L105

Sorry if it feels like a nitpick 😅

Reply:
ah sorry local var, I read instance variable in my head sorry. Sometimes I can't read properly >.<